### ISSUE TITLE
using dart:io's stdout stops fprintf(stdout, ... from working when called from C.

### ISSUE THREAD

**Issue by hterkelsen** [dart-lang/sdk#26872]
From the SDK directory
```
$ echo 'main() {}' > foo.dart

$ ./out/ReleaseX64/dart --print-metrics foo.dart
Printing metrics for foo.dart$main-759080858
heap.old.used 0.000 b (0)
heap.old.used.max 0.000 b (0)
heap.old.capacity 0.000 b (0)
heap.old.capacity.max 2620.000 kb (2682880)
heap.old.external 0.000 b (0)
heap.new.used 0.000 b (0)
heap.new.used.max 0.000 b (0)
heap.new.capacity 0.000 b (0)
heap.new.capacity.max 2048.000 kb (2097152)
heap.new.external 0.000 b (0)
heap.global.used 0.000 b (0)
heap.global.used.max -8.000 b (-8)

Printing metrics for VM
vm.isolate.count 0

Printing metrics for vm-isolate-416790917
heap.old.used 0.000 b (0)
heap.old.used.max 0.000 b (0)
heap.old.capacity 0.000 b (0)
heap.old.capacity.max 1876.000 kb (1921024)
heap.old.external 0.000 b (0)
heap.new.used 0.000 b (0)
heap.new.used.max 0.000 b (0)
heap.new.capacity 0.000 b (0)
heap.new.capacity.max 0.000 b (0)
heap.new.external 0.000 b (0)
heap.global.used 0.000 b (0)
heap.global.used.max 0.000 b (0)

$ ./out/ReleaseX64/dart --print-metrics --package-root=out/ReleaseX64/packages pkg/analyzer_cli/bin/analyzer.dart --dart-sdk=sdk foo.dart
Analyzing [foo.dart]...
No issues found
```

In the second case, no metrics are printed.

FYI, this works with dart 1.16.1, but not with dart 1.17.0 and beyond.

**Comment by johnmccutchan** [comment]
Thanks!

**Comment by hterkelsen** [comment]
It looks like it doesn't work with programs that call `exit`.

foo.dart:
```
import 'dart:io';

main() {
  exit(1);
}
```

```
$ dart --print-metrics foo.dart
```

**Comment by johnmccutchan** [comment]
@hterkelsen Yes, exit maps to C's exit which immediately exits the process.

**Comment by johnmccutchan** [comment]
And `--print-metrics` is run as part of VM shutdown (which is skipped if exit is called).

**Comment by johnmccutchan** [comment]
@hterkelsen Was the issue that your program was calling exit? If so, please close this issue.

**Comment by hterkelsen** [comment]
No, that was a red herring (the analyzer may call exit in some special cases, but not in the normal case). The program isn't calling exit, but is running until the end of `main` and terminating normally. Still there are no print metrics displayed.

**Comment by johnmccutchan** [comment]
Using tip of tree,

```
$ ./xcodebuild/DebugX64/dart --print-metrics ~/hw.dart
Hello World
Printing metrics for hw.dart$main-657318181
heap.old.used 0.000 b (0)
heap.old.used.max 0.000 b (0)
heap.old.capacity 0.000 b (0)
heap.old.capacity.max 2624.000 kb (2686976)
heap.old.external 0.000 b (0)
heap.new.used 0.000 b (0)
heap.new.used.max 0.000 b (0)
heap.new.capacity 0.000 b (0)
heap.new.capacity.max 2048.000 kb (2097152)
heap.new.external 0.000 b (0)
heap.global.used 0.000 b (0)
heap.global.used.max -6365935209750747136.000 b (-6365935209750747232)

Printing metrics for VM
vm.isolate.count 0

Printing metrics for vm-isolate-372721263
heap.old.used 0.000 b (0)
heap.old.used.max 0.000 b (0)
heap.old.capacity 0.000 b (0)
heap.old.capacity.max 1876.000 kb (1921024)
heap.old.external 0.000 b (0)
heap.new.used 0.000 b (0)
heap.new.used.max 0.000 b (0)
heap.new.capacity 0.000 b (0)
heap.new.capacity.max 0.000 b (0)
heap.new.external 0.000 b (0)
heap.global.used 0.000 b (0)
heap.global.used.max 0.000 b (0)
```

Where `hw.dart` is:

```dart
main() => print('Hello World');
```

**Comment by hterkelsen** [comment]
I can only reproduce the problem with the analyzer entrypoint. You should see it too if you do:

```
$ ./xcodebuild/DebugX64/dart --print-metrics ./pkg/analyzer_cli/bin/analyzer.dart ~/hw.dart
```

**Comment by johnmccutchan** [comment]
@hterkelsen Yes, I got to the same point. All programs I've tried but the `analyzer_cli` print their metrics. So this isn't a bug in the VM.

Digging in, I discovered the issue:

`analyzer_cli` is doing something with stdout and the metrics are printed on stdout. I did a local test where I switched to printing on stderr and I could see the output for the vm-isolate but no the analyzer isolate. You can see below.

```
$ ./xcodebuild/DebugX64/dart --print-metrics --package-root=xcodebuild/DebugX64/packages pkg/analyzer_cli/bin/analyzer.dart --dart-sdk=sdk ~/hw.dart
Analyzing [/Users/johnmccutchan/hw.dart]...
No issues found
Printing metrics for analyzer.dart$main-439201782
heap.old.used 0.000 b (0)
heap.old.used.max 19542.656 kb (20011680)
heap.old.capacity 0.000 b (0)
heap.old.capacity.max 59884.000 kb (61321216)
heap.old.external 0.000 b (0)
heap.new.used 0.000 b (0)
heap.new.used.max 8517.828 kb (8722256)
heap.new.capacity 0.000 b (0)
heap.new.capacity.max 32768.000 kb (33554432)
heap.new.external 0.000 b (0)
heap.global.used 0.000 b (0)
heap.global.used.max 65975.641 kb (67559056)

Printing metrics for VM
Printing metrics for vm-service
heap.old.used 0.000 b (0)
heap.old.used.max 0.000 b (0)
heap.old.capacity 0.000 b (0)
heap.old.capacity.max 5440.000 kb (5570560)
heap.old.external 0.000 b (0)
heap.new.used 0.000 b (0)
heap.new.used.max 92.719 kb (94944)
heap.new.capacity 0.000 b (0)
heap.new.capacity.max 2048.000 kb (2097152)
heap.new.external 0.000 b (0)
heap.global.used 0.000 b (0)
heap.global.used.max 4724.281 kb (4837664)

Printing metrics for vm-isolate-666753949
heap.old.used 0.000 b (0)
heap.old.used.max 0.000 b (0)
heap.old.capacity 0.000 b (0)
heap.old.capacity.max 1876.000 kb (1921024)
heap.old.external 0.000 b (0)
heap.new.used 0.000 b (0)
heap.new.used.max 0.000 b (0)
heap.new.capacity 0.000 b (0)
heap.new.capacity.max 0.000 b (0)
heap.new.external 0.000 b (0)
heap.global.used 0.000 b (0)
heap.global.used.max 0.000 b (0)
```

**Comment by hterkelsen** [comment]
foo.dart:

```dart
import 'dart:io';

main() {
  var so = stdout;
  so.writeln('hello world');
}
```

```
$ dart --print-metrics foo.dart
hello world
```

**Comment by johnmccutchan** [comment]
The example indicates there is a conflict between `dart:io`'s `stdout.writeln` and `fprintf(stdout, ...)` from C.

**Comment by whesse** [comment]
@sgjesse @whesse

**Comment by whesse** [comment]
@lrhn

**Comment by lrhn** [comment]
I don't see the problem on tip-of-tree on my local machine (x64 Linux), but it does show up on Windows 7 X64. What system was the above tests made on - the shell looks unix-y?

You don't even have to print on stdout, merely reading the `stdout` getter is enough to trigger the problem on Windows.

**Comment by johnmccutchan** [comment]
I was on a Mac

**Comment by hterkelsen** [comment]
I saw the problem on tip-of-tree on my Linux machine. FYI, --print-metrics is now printing on stderr instead, so I'm not sure how to reproduce the problem anymore, you'll have to revert [this CL][cl] if you want to repro with --print-metrics on the example above.

[cl]: https://codereview.chromium.org/2148613002/

**Comment by hterkelsen** [comment]
actually, you can see the problem on tip-of-tree with Linux if you run the VM with `--print-metrics` on a file that gets `stderr` from `dart:io`.

```dart
import 'dart:io';

main() {
  print(stderr.hashCode);
}
```

**Comment by floitschG** [comment]
The current implementation keeps track of all files (with finalizers and refcounts) and automatically closes them when it exits. `stdio` and `stderr` are treated like files and are therefore close at exit. (Similar to calling `stdio.close()` or `stderr.close()`.

Normally that isn't a problem, but if the embedder wants to use these handles after Dart has shut down, they are unusable...

We could just avoid automatically closing `stderr` and `stdio`.
In the following patch, `File.close()` doesn't depend on the destructor anymore to close files (the explicit `file->Close()` line), and the destructor wouldn't invoke `Close()` for stdout/stderr anymore, thus keeping it open for the embedder.

```patch
diff --git a/runtime/bin/file.cc b/runtime/bin/file.cc
index 43e1944..a3dcd7f 100644
--- a/runtime/bin/file.cc
+++ b/runtime/bin/file.cc
@@ -113,6 +113,7 @@ void FUNCTION_NAME(File_Exists)(Dart_NativeArguments args) {
 void FUNCTION_NAME(File_Close)(Dart_NativeArguments args) {
   File* file = GetFile(args);
   ASSERT(file != NULL);
+  file->Close();
   file->DeleteWeakHandle(Dart_CurrentIsolate());
   file->Release();
 
diff --git a/runtime/bin/file_linux.cc b/runtime/bin/file_linux.cc
index eab92d5..4f61db1 100644
--- a/runtime/bin/file_linux.cc
+++ b/runtime/bin/file_linux.cc
@@ -39,7 +39,8 @@ class FileHandle {
 
 
 File::~File() {
-  if (!IsClosed()) {
+  if (!IsClosed() &&
+      handle_->fd() != STDOUT_FILENO && handle_->fd() != STDERR_FILENO) {
     Close();
   }
   delete handle_;
```

If the user invokes `stderr.close()` explicitly, the `--print-metrics` would still go into the void, but at least there would be a good explanation for it.

Wdyt?

**Comment by sgjesse** [comment]
This approach looks good to me. We could even consider not closing any files automatically, and leave that part to standard stdio behaviour.

**Comment by johnmccutchan** [comment]
@floitschG SGTM

**Comment by floitschG** [comment]
Soeren and I discussed this shortly in person: it's probably not a good idea to completely stop closing files when they are GCed. It's a bad idea to rely on the GC to do work for you, but since we implemented it this way and shipped the IO library a long time ago, we shouldn't just break them (in probably very difficult-to-debug ways).

However, I will disable the automatic closing of stdout and stderr.
